### PR TITLE
compose (dev): don't delete stored data when starting

### PR DESCRIPTION
**Pull request by vitoravelino**

### CODE REVIEW

**Comment by vitoravelino** [comment]
@mssola 

I tried to mimic what you did with `background.rb` and it's working but a weird error is happening with the background container when database is not empty.

Sometimes it happens a `Connection reset by peer`, which is fine because the current rescue catches the error and doesn't break the process, but sometimes an `EOFError` error is happening and causing the container to exit.

```
background_1  | bin/rails: end of file reached (EOFError)
portus_background_1 exited with code 1
```

I added `StandardError` to the rescue on the `background.rb` and I got

```
#<EOFError: end of file reached>
Exception: end of file reached
```

There's no backtrace for that. :(

Catching the `StandardError` solves the problem but...

File: examples/development/compose/init.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    end
+    exit 1
+  end
+
+  Rails.logger.tagged("Database") { Rails.logger.error "Not ready yet. Waiting..." }
+  sleep 5
+  count += 5
+
+  system("bundle exec rake db:setup") if status == :missing
+end
```
Comment (mssola) [inline examples/development/compose/init.rb]: You might want to put this code in `lib/portus/db` as a method, and then use this both here and in the background.rb file 😉

**Comment by vitoravelino** [comment]
> This is weird... Maybe you can rescue this exception and print with the #message or #backtrace methods of said exception.

As I said, no backtrace and message is what I pasted. I'll double check anyway.

File: examples/development/compose/init.rb (hunk 1 above)
Comment (vitoravelino) [reply examples/development/compose/init.rb]: And be able to pass a block to it? That sounds interesting...

**Comment by mssola** [comment]
Can you tell what's the status of the DB then ? Because if it's just fine, maybe we can catch this exception, log it and simply ignore it 🤔

**Comment by vitoravelino** [comment]
> Can you tell what's the status of the DB then ? Because if it's just fine, maybe we can catch this exception, log it and simply ignore it 🤔

I'll reproduce here and give you the info about the db.

I would be fine doing what you suggested since apparently it doesn't change the flow or create a side effect.

**Comment by vitoravelino** [comment]
I totally forgot to context where this error comes from. It happens inside the `sync.rb` when fetching the catalog. Here's the backtrace:

```
<internal:prelude>:78:in `__read_nonblock'
<internal:prelude>:78:in `read_nonblock'
/usr/lib64/ruby/2.4.0/net/protocol.rb:172:in `rbuf_fill'
/usr/lib64/ruby/2.4.0/net/protocol.rb:154:in `readuntil'
/usr/lib64/ruby/2.4.0/net/protocol.rb:164:in `readline'
/usr/lib64/ruby/2.4.0/net/http/response.rb:40:in `read_status_line'
/usr/lib64/ruby/2.4.0/net/http/response.rb:29:in `read_new'
/usr/lib64/ruby/2.4.0/net/http.rb:1446:in `block in transport_request'
/usr/lib64/ruby/2.4.0/net/http.rb:1443:in `catch'
/usr/lib64/ruby/2.4.0/net/http.rb:1443:in `transport_request'
/usr/lib64/ruby/2.4.0/net/http.rb:1416:in `request'
/srv/Portus/lib/portus/http_helpers.rb:159:in `block in get_response_token'
/usr/lib64/ruby/2.4.0/net/http.rb:877:in `start'
/usr/lib64/ruby/2.4.0/net/http.rb:608:in `start'
/srv/Portus/lib/portus/http_helpers.rb:158:in `get_response_token'
/srv/Portus/lib/portus/http_helpers.rb:115:in `request_auth_token'
/srv/Portus/lib/portus/http_helpers.rb:48:in `perform_request'
/srv/Portus/lib/portus/registry_client.rb:128:in `get_page'
/srv/Portus/lib/portus/registry_client.rb:115:in `paged_response'
/srv/Portus/lib/portus/registry_client.rb:80:in `catalog'
/srv/Portus/lib/portus/background/sync.rb:28:in `block in execute!'
/usr/lib64/ruby/gems/2.4.0/gems/activerecord-4.2.10/lib/active_record/relation/batches.rb:51:in `block (2 levels) in find_each'
/usr/lib64/ruby/gems/2.4.0/gems/activerecord-4.2.10/lib/active_record/relation/batches.rb:51:in `each'
/usr/lib64/ruby/gems/2.4.0/gems/activerecord-4.2.10/lib/active_record/relation/batches.rb:51:in `block in find_each'
/usr/lib64/ruby/gems/2.4.0/gems/activerecord-4.2.10/lib/active_record/relation/batches.rb:124:in `find_in_batches'
/usr/lib64/ruby/gems/2.4.0/gems/activerecord-4.2.10/lib/active_record/relation/batches.rb:50:in `find_each'
/usr/lib64/ruby/gems/2.4.0/gems/activerecord-4.2.10/lib/active_record/querying.rb:9:in `find_each'
/srv/Portus/lib/portus/background/sync.rb:25:in `execute!'
/srv/Portus/bin/background.rb:74:in `block (2 levels) in <top (required)>'
/srv/Portus/bin/background.rb:69:in `each'
/srv/Portus/bin/background.rb:69:in `block in <top (required)>'
/srv/Portus/bin/background.rb:68:in `loop'
/srv/Portus/bin/background.rb:68:in `<top (required)>'
/usr/lib64/ruby/gems/2.4.0/gems/railties-4.2.10/lib/rails/commands/runner.rb:60:in `load'
/usr/lib64/ruby/gems/2.4.0/gems/railties-4.2.10/lib/rails/commands/runner.rb:60:in `<top (required)>'
/usr/lib64/ruby/gems/2.4.0/gems/railties-4.2.10/lib/rails/commands/commands_tasks.rb:123:in `require'
/usr/lib64/ruby/gems/2.4.0/gems/railties-4.2.10/lib/rails/commands/commands_tasks.rb:123:in `require_command!'
/usr/lib64/ruby/gems/2.4.0/gems/railties-4.2.10/lib/rails/commands/commands_tasks.rb:90:in `runner'
/usr/lib64/ruby/gems/2.4.0/gems/railties-4.2.10/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
/usr/lib64/ruby/gems/2.4.0/gems/railties-4.2.10/lib/rails/commands.rb:17:in `<top (required)>'
bin/rails:12:in `require'
bin/rails:12:in `<main>'
```

And the message is:

`
bin/rails: end of file reached (EOFError)
`

Any idea?

**Comment by vitoravelino** [comment]
While I was testing my refactor, I was able to notice that when puma is not ready 2 exceptions keep happening (`connection reset by peer` and `end of file reached`). So I guess this error is something expected in some situations from HTTP.

I'll write the test and the doc when I get your +1.

File: lib/portus/errors.rb [hunk 2]
```diff
@@ -37,5 +37,8 @@ class NotFoundError < RuntimeError; end
 
     # Raised if this client does not have the credentials to perform an API call.
     class CredentialsMissingError < RuntimeError; end
+
+    # Raised if any timeout reached
+    class TimeoutReachedError < RuntimeError; end
```
Comment (mssola) [inline lib/portus/errors.rb]: Be careful, these errors are for registry-related exceptions, and this one right here is for the DB. I would move it to Portus::DB.

Comment (vitoravelino) [reply lib/portus/errors.rb]: Ok, I'll move it then. No worries. :)

Comment (vitoravelino) [reply lib/portus/errors.rb]: Done

**Comment by mssola** [comment]
@vitoravelino are you planning to provide tests for those exceptions on this PR, or in another one ? Just saying because it's still marked as wip but it looks good to me so far 😄

**Comment by vitoravelino** [comment]
> @vitoravelino are you planning to provide tests for those exceptions on this PR, or in another one ? Just saying because it's still marked as wip but it looks good to me so far 😄

I wasn't planning to provide tests for the `EOFError`, just for the new `wait_until` method to fix the coverage. I'm working on it at the moment.

**Comment by vitoravelino** [comment]
@mssola good to go!

**Comment by mssola** [comment]
Thanks 👏

**Pull request merged**